src/bt-hh-io.c

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (C) 2014-2015  Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <assert.h>
+#include <fdio/task.h>
+#include <hardware/bluetooth.h>
+#include <hardware/bt_hh.h>
+#include <stdlib.h>
+#include "compiler.h"
+#include "log.h"
+#include "bt-proto.h"
+#include "bt-pdubuf.h"
+#include "bt-core-io.h"
+#include "bt-hh-io.h"
+
+
+enum {
+/* commands/responses */
+OPCODE_CONNECT = 0x01,
+OPCODE_DISCONNECT = 0x02,
+OPCODE_VIRTUAL_UNPLUG = 0x03,
+OPCODE_SET_INFO = 0x04,
+OPCODE_GET_PROTOCOL = 0x05,
+OPCODE_SET_PROTOCOL = 0x06,
+OPCODE_GET_REPORT = 0x07,
+OPCODE_SET_REPORT = 0x08,
+OPCODE_SEND_DATA = 0x09,
+/* notifications */
+OPCODE_CONNECTION_STATE_NTF = 0x81,
+OPCODE_HID_INFO_NTF = 0x82,
+OPCODE_PROTOCOL_MODE_NTF = 0x83,
+OPCODE_IDLE_TIME_NTF = 0x84,
+OPCODE_GET_REPORT_NTF = 0x85,
+OPCODE_VIRTUAL_UNPLUG_NTF = 0x86,
+OPCODE_HANDSHAKE_NTF = 0x87
```

Comment:
This notification is only present on L and later. Please protect by '#if ANDROID_VERSION >= 21'